improve LLVM IR typing and fix the OOM issue reported in #528

mmarchini opened the pull request:
(#667 needs to land first. Changes in this PR are on [this commit](https://github.com/iovisor/bpftrace/commit/bc3d16ea6b5bc155118180b1ebfe0c74366c70db))

I was trying to find a way to fix #528, which has been very challenging since the OOM crash happens after IR generation (i.e., during optimization or BPF generation, when we're entirely in LLVM context), and it occurred to me that the issue could be happening because of somethings we're doing wrong in our IR generation. LLVM has **a lot** of aserts to ensure correctness of the IR generation in their debug build (unfortunately most of those checks are not performed in their release build), so I
did a Debug build of LLVM (6) and used it to find typing issues in our IR generation. I started with the code from #667, since @michalgr already fixed several assert errors. I found a few more while trying to run the examples from #528, and after fixing those the OOM issue seems to be fixed (for the scenarios we found at least).

While working on this, I also found a couple of things we can improve in our IR generation:

- Don't promote every integer to 64-bits by making sure we properly check and propagate their types.
- For known structs and arrays, use StructLayout + GEP instead of pointer arithmetic

mmarchini commented:
(btw, I'll need to update codegen tests before merging this)

ajor commented:
This script prints "1234" instead of "0" when `CreateAllocaBPFInit` is changed to `CreateAllocaBPF`:
```
BEGIN
{
  @x = 0;
  if (@x == 1)
  {
    $var = 1234;
  }
  printf("%d\n", $var);
}
```

mmarchini commented:
> This script prints "1234" instead of "0" when CreateAllocaBPFInit is changed to CreateAllocaBPF:

Turns out changing to `CreateAllocaBPF` was not necessary (I don't remember why I did that).

I've rebased and added runtime tests for the LLVM OOM issue + https://github.com/iovisor/bpftrace/pull/711#issuecomment-505097671. Will wait some LGTMs before updating the codegen tests, since almost all of them changed.

danobi commented on src/ast/codegen_llvm.cpp:
```diff
@@ -991,7 +991,12 @@ void CodegenLLVM::visit(FieldAccess &acc)
-    Value *src = b_.CreateAdd(expr_, b_.getInt64(field.offset));
+    // LLVM doesn't like Binops with different types, so we convert our pointer
+    // to integer and then back to pointer.
+    // Value *src = b_.CreateAdd(expr_, b_.getInt64(field.offset));
```
Maybe to make it more clear for future readers:
```suggestion
    // Instead of doing:
    //     Value *src = b_.CreateAdd(expr_, b_.getInt64(field.offset));
    // we do the below:
```

danobi commented on src/ast/codegen_llvm.cpp:
```diff
@@ -991,7 +991,12 @@ void CodegenLLVM::visit(FieldAccess &acc)
     // so expr_ will contain an external pointer to the start of the struct
 
-    Value *src = b_.CreateAdd(expr_, b_.getInt64(field.offset));
+    // LLVM doesn't like Binops with different types, so we convert our pointer
```
Can you clarify what "doesn't like" means in the comment?

danobi commented on src/ast/codegen_llvm.cpp:
```diff
@@ -1043,7 +1048,14 @@ void CodegenLLVM::visit(ArrayAccess &arr)
 
   AllocaInst *dst = b_.CreateAllocaBPF(SizedType(Type::integer, type.pointee_size), "array_access");
-  Value *src = b_.CreateAdd(array, offset);
+  // Value *src = b_.CreateAdd(array, offset);
```
Same as above so it doesn't look like a debugging comment

danobi commented on src/ast/codegen_llvm.cpp:
```diff
@@ -1672,7 +1684,8 @@ void CodegenLLVM::createFormatStringCall(Call &call, int &id, CallArgs &call_arg
     else
-      b_.CreateStore(expr_, offset);
+      // Promote to 64-bits to match `offset` type
+      b_.CreateStore(b_.CreateIntCast(expr_, b_.getInt64Ty(), false), offset);
```
`Field::offset` is type `int`. That would be a 32-bit int right? Maybe we should make `Field::offset` `uint64_t` or something.

danobi commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -69,16 +69,18 @@ AllocaInst *IRBuilderBPF::CreateAllocaBPFInit(const SizedType &stype, const std:
   else
   {
-    CreateMemSet(alloca, getInt64(0), stype.size, 1);
+    alloca = CreateAllocaBPF(getIntPtrTy(module_.getDataLayout()), nullptr, name);
```
So this branch allocates an `int*` on the stack. Where does the pointed-to array get allocated? Is the memset just gonna scribble all over the stack?

danobi commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -265,7 +267,7 @@ void IRBuilderBPF::CreateProbeRead(AllocaInst *dst, size_t size, Value *src)
   FunctionType *proberead_func_type = FunctionType::get(
       getInt64Ty(),
-      {getInt8PtrTy(), getInt64Ty(), getInt8PtrTy()},
+      {dst->getType(), getInt64Ty(), src->getType()},
```
Isn't the second arg a u32? Not a int64?

https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L555

danobi commented:
Closing due to inactivity. Please reopen if necessary.

I haven't really seen the OOM issue recently
